lib/client/db/postgres/repl/repl.go

```diff
@@ -0,0 +1,258 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package repl
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"strings"
+
+	"github.com/gravitational/trace"
+	"github.com/jackc/pgconn"
+	"golang.org/x/term"
+
+	"github.com/gravitational/teleport"
+	clientproto "github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/lib/asciitable"
+)
+
+type REPL struct {
+	ctx        context.Context
+	cancelFunc context.CancelCauseFunc
+	conn       *pgconn.PgConn
+	client     io.ReadWriter
+	serverConn net.Conn
+	route      clientproto.RouteToDatabase
+	term       *term.Terminal
+	commands   map[string]*command
+}
+
+func Start(ctx context.Context, client io.ReadWriter, serverConn net.Conn, route clientproto.RouteToDatabase) (*REPL, error) {
+	config, err := pgconn.ParseConfig(fmt.Sprintf("postgres://%s@%s/%s", route.Username, hostnamePlaceholder, route.Database))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	config.TLSConfig = nil
+
+	// Provide a lookup function to avoid having the hostname placeholder to
+	// resolve into something else. Note that the returned value won't be used.
+	config.LookupFunc = func(_ context.Context, _ string) ([]string, error) {
+		return []string{hostnamePlaceholder}, nil
+	}
+	config.DialFunc = func(_ context.Context, _, _ string) (net.Conn, error) {
+		return serverConn, nil
+	}
+
+	pgConn, err := pgconn.ConnectConfig(ctx, config)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	replCtx, cancelFunc := context.WithCancelCause(ctx)
+	r := &REPL{
+		ctx:        replCtx,
+		cancelFunc: cancelFunc,
+		conn:       pgConn,
+		client:     client,
+		serverConn: serverConn,
+		route:      route,
+		term:       term.NewTerminal(client, ""),
+		commands:   initCommands(),
+	}
+
+	go r.start()
+	return r, nil
+}
+
+func (r *REPL) Close() {
+	r.close(nil)
+}
+
+func (r *REPL) close(err error) {
+	r.cancelFunc(err)
+	r.conn.Close(r.ctx)
```

Comment:
This will call close with a context that's already canceled, won't it?

Reply:
I've updated the structure of the REPL to use only two functions `New` and `Run(ctx)`, so that the REPL instance doesn't require holding the context reference or implementing closing.